### PR TITLE
include: Move the internal headers to src.

### PR DESCRIPTION
**Pull request by AlexanderKurtz**
This means that "the public API" and "what is in the
'include' directory" are now finally the same thing.

### CODE REVIEW

**Comment by AlexanderKurtz** [comment]
Hm, Travis failed... Investigating...

**Comment by AlexanderKurtz** [comment]
It seems I forgot about a few Makefiles in #102, sorry about that. Should work now!

**Comment by weidendo** [comment]
Now I am confused.

We want a LAIK installation to enable LAIK plugin development, such as agents, backends, etc.
I assume that the moved headers are not installed any longer? I can imagine that plugin
developers need access to structs sometimes.

**Comment by AlexanderKurtz** [comment]
I figure the fixup for #102 shouldn't be blocked by the discussion here, so I opened up #106 for this!

**Comment by AlexanderKurtz** [comment]
> We want a LAIK installation to enable LAIK plugin development, such as agents, backends, etc.
> I assume that the moved headers are not installed any longer? I can imagine that plugin
> developers need access to structs sometimes.

The agent interface (```agent.h``` and ```ext.h```) **is** part of the public API (included by ```laik.h```) and remains in ```include```. All the existing plugins (i.e. everything under ```external```) only include ```laik.h``` and not ```laik-internal.h```.

Yes, it is possible that a future plugin may want to access something from ```laik-internal.h``` and the (existing) backends also have to do this, but this doesn't mean that ```laik-internal.h``` is part of the public API.

In fact, looking at commits like f538bd624e70132c3a2520edd76b33db3d9db6ca and acf9d0e0f293d4de40e9de2db3108ea18cf9b210 suggests that ```laik-internal.h``` **is not stable at all** and should consequently not be marked as such (i.e. it should not be part of ```include```).

TL;DR: Things only accessing ```laik.h``` can realistically  be developed outside of LAIK itself, while things using ```laik-internal.h``` should really live inside the main LAIK repo. Do you agree?

**Comment by weidendo** [comment]
Hmm. Nothing is stable at the moment.

Communication backends should be able to be written as plugins.
Is the TCP backend fine with just including "laik.h"?

**Comment by AlexanderKurtz** [comment]
> Communication backends should be able to be written as plugins.
> Is the TCP backend fine with just including "laik.h"?

No backend (not even the ```single``` backend) works with only ```laik.h```. Using some [iwyu](https://github.com/include-what-you-use/include-what-you-use/) magic, we can determine what exactly is used:

# Single backend

```
#include <laik-internal.h>  // for _Laik_Transition, _Laik_MappingList, laik_isInGroup, _Laik_Data, _Laik_Instance, _Laik_Mapping, redTOp, _Laik_Space, Laik_TransitionPlan, _Laik_Backend
#include <laik.h>           // for Laik_MappingList, Laik_Instance, Laik_Transition, Laik_Data, Laik_Mapping, laik_log, Laik_Space, laik_new_instance, Laik_Group, Laik_Index, Laik_Slice, Laik_Backend
```

# MPI backend

```
#include <laik-internal.h>  // for redTOp, _Laik_Group, TaskGroup, _Laik_Mapping, _Laik_Data, recvTOp, Laik_TransitionPlan, _Laik_Transition, sendTOp, _Laik_MappingList, laik_isInGroup, _Laik_Layout, _Laik_SwitchStat, laik_allocateMap, _Laik_Type, _Laik_Instance, _Laik_Backend, _Laik_Partitioning, _Laik_Space
#include <laik.h>           // for Laik_Group, Laik_Mapping, laik_log, Laik_Data, Laik_Transition, laik_log_append, Laik_Slice, Laik_Index, Laik_MappingList, laik_log_begin, laik_log_flush, laik_log_Slice, laik_panic, laik_slice_size, Laik_Instance, Laik_SwitchStat, Laik_Layout, laik_index_isEqual, Laik_Type, laik_new_instance, _Laik_LogLevel::LAIK_LL_Panic, Laik_Backend, Laik_Partitioning, Laik_Space, _Laik_ReductionOperation::LAIK_RO_Sum, laik_Double, laik_Float
```

# New TCP backend

```
#include <laik-internal.h>  // for _Laik_Data, _Laik_Group, redTOp, _Laik_Mapping, Laik_TransitionPlan, TaskGroup, _Laik_MappingList, _Laik_Transition, recvTOp, _Laik_Space, sendTOp, _Laik_Layout, _Laik_SwitchStat, laik_allocateMap, _Laik_Partitioning, _Laik_Type, _Laik_Backend, _Laik_Instance
#include <laik.h>           // for Laik_Data, Laik_Group, Laik_Mapping, Laik_Slice, Laik_Index, Laik_Transition, Laik_MappingList, laik_slice_size, Laik_Space, Laik_Layout, Laik_SwitchStat, laik_index_isEqual, Laik_Partitioning, Laik_Instance, Laik_Type, laik_new_instance, Laik_Backend, Laik_ReductionOperation, _Laik_ReductionOperation::LAIK_RO_Sum, laik_Double, laik_Float
```

So, if we want to write backends as plugins, we either need to add (a lot of) getters and setters or declare ```laik-internal.h``` part of the API. Since I don't think that either alternative is realistic, backends shouldn't be plugins IMHO.

**Comment by AlexanderKurtz** [comment]
This is obsoleted by #113!

**Comment by weidendo** [comment]
> So, if we want to write backends as plugins...

This is purely a strategic issue, and technically, this always can be done if wanted.

**Comment by AlexanderKurtz** [comment]
Sure! The whole problem I have with the current situation is that the name ```laik-internal.h``` suggests (at least to me) that this isn't part of the API, but "internal" stuff...

**Comment by weidendo** [comment]
Before we do not have a single backend which enables LAIK applications to become as fast as MPI applications, we have no idea how the best backend API looks like. Thus, it is mood to discuss about splitting headers for plugins, as well as about the best place for different types of headers.

Can we just leave it this way at the moment?
